editor/src/components/images.spec.ts

```diff
@@ -0,0 +1,136 @@
+import { parseMultiplier, FilenameParts, filenameFromParts, getFilenameParts } from './images'
+
+describe('image related helper functions', () => {
+  it('parseMultiplier', () => {
+    const inputs: Array<[string, number | null]> = [
+      ['2x', 2],
+      ['3x', 3],
+      ['22x', 22],
+      ['0x', 0],
+      ['3241x', 3241],
+      ['0xxx', null],
+      ['3241_x', null],
+      ['32_41x', null],
+    ]
+
+    for (const [input, result] of inputs) {
+      expect(parseMultiplier(input)).toEqual(result)
+    }
+  })
+
+  it('print filename object', () => {
+    const filenames: Array<[FilenameParts, string]> = [
+      [{ filename: 'stuff', extension: 'png' }, 'stuff.png'],
+      [{ filename: 'stuff', extension: 'png', multiplier: 2 }, 'stuff@2x.png'],
+      [
+        { filename: 'stuff', extension: 'png', multiplier: 2, deduplicationSeqNumber: 2 },
+        'stuff_2@2x.png',
+      ],
+    ]
+
+    for (const [input, result] of filenames) {
+      const parsed = filenameFromParts(input)
+      expect(parsed).toEqual(result)
+    }
+  })
+
+  it('parse image filenames', () => {
+    const filenames: Array<[FilenameParts, string]> = [
+      [{ filename: 'stuff', extension: 'png' }, 'stuff.png'],
+      [{ filename: 'stuff', extension: 'png', multiplier: 2 }, 'stuff@2x.png'],
+      [
+        { filename: 'stuff', extension: 'png', multiplier: 2, deduplicationSeqNumber: 2 },
+        'stuff_2@2x.png',
+      ],
+      [
+        { filename: 'stuff@3', extension: 'png', multiplier: 2, deduplicationSeqNumber: 2 },
+        'stuff@3_2@2x.png',
+      ],
+    ]
```

Comment:
nit: for these tests, I think it'd read much better if you'd flip the input array to be [[input, result]] !

Reply:
absolutely, fixed!